### ISSUE TITLE
Switch to Webp

### ISSUE THREAD

**Issue by oong819** [minetest/minetest#12620]
## Problem
Reduce app size so more people can install
## Solutions
Convert all image to webp in lossless mode to reduce app size

## Alternatives
idk

## Additional context

i dont know is the irrlichMt engine suppot webp or not

**Comment by rubenwardy** [comment]
Hi, thanks for the suggestion.

Minetest is currently 18MB to download, and media doesn't go above 200MB when playing. This is already much much smaller than a lot of modern 3d games, they can reach into GBs. It's also worth noting that most of the image files are 16x16, so the techniques that more efficient formats use may not actually be better at this size.

I don't really have much of an opinion on using webp, but it's extra work and formats like [.glTF](https://github.com/minetest/minetest/issues/9673) should be prioritised

**Comment by erlehmann** [comment]
> Convert all image to webp in lossless mode to reduce app size

I know you are talking about lossless WebP mode here, but you should know that WebP propaganda is highly misleading, e.g. intentionally comparing JPEG performance against cjpeg (a reference encoder IIRC) instead of the highly optimized mozjpeg.

Only in such a cherry-picked benchmark WebP is much better than JPEG: https://siipo.la/blog/is-webp-really-better-than-jpeg

Unless you provide benchmarks, I would assume that the claims about WebP being better than PNG are similarly misleading.

> It's also worth noting that most of the image files are 16x16, so the techniques that more efficient formats use may not actually be better at this size.

At tiny sizes even uncompressed bitmaps can be competitive, especially when served in a zip file (which Minetest content is).
I have made a shell script with which you can verify that claim: https://forum.minetest.net/viewtopic.php?p=412499#p412499

The gains of any texture compression improvement are so small that it does not matter. Also transfer size is not the same as apparent byte size.

Example: Imagine you are making a 16×16 texture pack. By using uncompressed TGA files you could make the zip download 5% to 15% smaller than using an optimized PNG file. But that will only turn a download of 85 kilobytes into a download of 80 kilobytes – or a download that is 10.6 megabytes into a download that is 10.3 megabytes. Given that the files are downloaded very rarely, it does not seem to be worth it for most people to think about these optimizations.

If Minetest had a lot of much bigger textures and PNG encoders were generally bad, you might have a point. For example, devtest generates a 512×512 checkerboard texture and saves it as a PNG file in a non-optimal way. Using optipng on that file reduces its size by 95%. If a new graphicss format provided that kind of filesize improvement, people would consider it. But if a bitmap file is very small and only a few hundred bytes to begin with, even the best compression will usually shave off only tens of bytes.

TL;DR: No benchmarks, no proof of concept, no use case. Additionally, WebP has been discussed on IRC previously – this is how I learned of the misleading claims regarding WebP vs JPEG and the cherry-picked benchmarks. I believe this issue can be closed.

**Comment by erlehmann** [comment]
> cache media maybe gets as much as 200MB when playing.

Highly dependent on what servers you join. I have a >500MB cache folder from 2.5 years of occasional play.

Better per-file compression would not improve things: Most files are so small they use only one disk block (4k).

**Comment by oong819** [comment]
i will try it and will tell the result

**Comment by Montandalar** [comment]
At best, WebP would reduce transfer but as erlehmann pointed out it's not benchmarked very well and would probably actually reduce performance for little gain, and as we support mobile platforms, the performance hit is not worth it. Some people's phones can barely run the game as it is. Anyway, sorry to beat a dead horse, but in case anyone comes a-knocking to try to revive it..

**Comment by NA0341** [comment]
@erlehmann said:
> Highly dependent on what servers you join. I have a >500MB cache folder from 2.5 years of occasional play.

Yup that's pretty normal.
I joined about 5 to 7 Servers with "some mods installed" (TechAge and smaller)
~ and my `~/.cache/minetest/media/` folder currently contains 62.420 items with a (storage) size of 748,3MB.

---

Since I'm talking about this stuff, let me share some personal experience:

**Tip for Players who play with Multiple Accounts or on Multiple Servers** (GNU/Linux)

Okay it's actually a Guide.
And since it became too long, you can view it [here])https://share.note.sx/bxvofior)
(There's even more ~ and it gets updated and extended whenever I have time to do so. So make sure to save that page as a bookmark ;)

**Comment by oong819** [comment]
(this dumb request have been made by me 2 years ago, those time i'm pretty new to this)

**Comment by NA0341** [comment]
@oong819 No need to blame yourself =)
WebP can indeed make things a lot smaller with same perceptive image quality ~ in some situations.

But in Minetest's case where there are very simple graphics, optimized PNG's are a much better option.
WebP is also not just a new format, but the beginning of the Video Codec Compression Masterrace. WebP's compression is basically the same as the compression types of vp8 and vp9 Video Codecs. And folowed by AV1 we have a potentially even »more efficient« image codec. But whether it's any good or not is another thing (because an image that doesn't support any kind of metadata isn't all that useful ~ missing application support is another thing to consider).

After all, HEIC or HEIF (High Efficiency Image Codec (Format)), which is based on HEVC (High Efficiency Video Codec), also known as H.265 (standard JPEG is H.264), could still be considered an improvement to JPEG since it is almost drop-in as long as the application can read the codec.

For reference, check out [OptiPNG ~ C](https://optipng.sourceforge.net) or it's rewrite [OxiPNG ~ Rust](https://github.com/shssoichiro/oxipng) & [Curtail](https://github.com/Huluti/Curtail) ~ which are amazing tools for loss(y)less image compression.

**Comment by oong819** [comment]
> @oong819 No need to blame yourself =) WebP can indeed make things a lot smaller with same perceptive image quality ~ in some situations. (For use in Comic Book files, for example. Their image content works pretty well with WebP's compression.)
> 
> But in Minetest's case where there are very simple graphics, optimized PNG's are a much better option. WebP is also not just a new format, but the beginning of the Video Codec Compression Masterrace. WebP's compression is basically the same as the compression types of vp8 and vp9 Video Codecs. And folowed by AV1 we have a potentially even »more efficient« image codec. But whether it's any good or not is another thing (because an image that doesn't support any kind of metadata isn't all that useful ~ missing application support is another thing to consider).
> 
> After all, HEIC or HEIF (High Efficiency Image Codec (Format)), which is based on HEVC (High Efficiency Video Codec), also known as H.265 (standard JPEG is H.264), could still be considered an improvement to JPEG since it is almost drop-in as long as the application can read the codec.
> 
> For reference, check out [OptiPNG (C)](https://optipng.sourceforge.net) or it's rewrite [OxiPNG (Rust)](https://github.com/shssoichiro/oxipng) & [Curtail](https://github.com/Huluti/Curtail) ~ which are amazing tools for loss(y)less image compression.

I missed so many points in 2022, PNG is more simple and decompress faster than WebP, better fit for real-time rendering. And then we have security, compatibility, license,..........

Now (first half of 2024) i undestand how OptiPNG works, taken a look inside PNG doc and even tried to code a PNG decoder in Java (which is still unfinished today)

**Comment by NA0341** [comment]
As you may have read, OptiPNG also improves decoding performance & speed. (Besides improving compression & data structure (integrity).)

**Comment by appgurueu** [comment]
I love PNG. It's simple enough that you can write a short decoder in pure Lua, and an inefficient encoder is also easy to make. See https://github.com/appgurueu/modlib/blob/5f0dea2780b88d44d85b9704e0e81348c459404d/minetest/png.lua.
